### ISSUE TITLE
Reproduce `Pointer-Generator + Coverage` summarization model with single gpu

### ISSUE THREAD

**Issue by yuhui-zh15** [OpenNMT/OpenNMT-py#1610]
Hello,

Thank you for developing this tool! Recently, I'm trying to reproduce `Pointer-Generator + Coverage` model mentioned in [OpenNMT-py doc](http://opennmt.net/OpenNMT-py/Summarization.html).

I found if I use the default multi-gpu settings, the training would be stuck after thousands of iterations. Then I tried to use the single gpu with 2x default batch size to reproduce the result. However, my result is 1.5 points below the reported results. Could you let me know how to reproduce the same result with single gpu? 

Best,
Yuhui

**Comment by yuhui-zh15** [comment]
The training script I used:

**Comment by yuhui-zh15** [comment]
Here are the results I got:

```
---------------------------------------------
1 ROUGE-1 Average_R: 0.40794 (95%-conf.int. 0.40541 - 0.41041)
1 ROUGE-1 Average_P: 0.38831 (95%-conf.int. 0.38562 - 0.39113)
1 ROUGE-1 Average_F: 0.38287 (95%-conf.int. 0.38077 - 0.38503)
---------------------------------------------
1 ROUGE-2 Average_R: 0.17279 (95%-conf.int. 0.17060 - 0.17493)
1 ROUGE-2 Average_P: 0.16881 (95%-conf.int. 0.16636 - 0.17131)
1 ROUGE-2 Average_F: 0.16407 (95%-conf.int. 0.16202 - 0.16634)
---------------------------------------------
1 ROUGE-L Average_R: 0.37425 (95%-conf.int. 0.37189 - 0.37656)
1 ROUGE-L Average_P: 0.35701 (95%-conf.int. 0.35440 - 0.35984)
1 ROUGE-L Average_F: 0.35165 (95%-conf.int. 0.34959 - 0.35382)

* method sent_tag_verbatim
['rouge_1_recall', 'rouge_1_precision', 'rouge_1_f_score', 'rouge_2_recall', 'rouge_2_precision', 'rouge_2_f_score', 'rouge_l_recall', 'rouge_l_precision', 'rouge_l_f_score']
40.79   38.83   38.29   17.28   16.88   16.41   37.43   35.70   35.16
* evaluated 11490 samples, took 264.493s, averaging 0.023s/sample
* portion of samples that contains self-repetitions
full-sent,32-gram,16-gram,8-gram,4-gram,2-gram
0.00%,  0.00%,  0.00%,  0.00%,  0.00%,  40.88%,
* evaluated 11490 samples, took 1.653s, averaging 0.000s/sample
```

Here are the results reported in the website:

```
Model Type  | R1 R | R1 P | R1 F | R2 R | R2 P | R2 F | RL R | RL P | RL F
Pointer-Generator + Coverage | 39.05 | 43.02 | 39.53 | 17.16 | 18.77 | 17.28 | 35.98 | 39.56 | 36.38
```

**Comment by pltrdy** [comment]
First, Pointer-Generator results are from https://github.com/abisee/pointer-generator/ . The implementation is not exactly the same in OpenNMT-py, so technically, you're trying to reproduce the OpenNMT BRNN (1 layer, emb 128, hid 512), results are similar, was just to clarify.

Note that, 1 iteration is `n_gpu` GPU processing `acc` batch (here `acc=1`), therefore, using 2 GPUs will make iterations "bigger" i.e. 2 times more data processed. Therefore, using 1 GPU instead of 2, you should set `-accum_count 2` to be exactly comparable. 

In general you can change the batch size, the accum, and the GPU but keep the same final value of `batch_size x accum_count x n_gpu`.

**Comment by yuhui-zh15** [comment]
Thanks a lot for your reply. 

However, I still can not reproduce the result with either `-batch_size=2*16` or  `-accum_count 2`. All the models are trained 200k iterations, and I picked the best model based on PPL. I carefully checked all the arguments mentioned in the [webpage](http://opennmt.net/OpenNMT-py/Summarization.html) and [paper](https://arxiv.org/abs/1808.10792) `Bottom-up Summarization`, but there is still a gap between my reproduced result and the paper result. Do you have any idea how to reproduce the paper?

```
accum_count=2, batch_size=16: R1=38.24, R2=16.42, RL=35.15
accum_count=1, batch_size=32: R1=38.29, R2=16.41, RL=35.16
accum_count=1, batch_size=16: R1=38.24, R2=16.42, RL=35.17
paper (Pointer-Generator + Coverage Penalty): R1=39.12, R2=17.35, RL=36.12
```

**Comment by pltrdy** [comment]
It's not trivial for sure, we've been discussing it for some time. 

Check that you are correctly evaluating (with `rouge-baseline` or `files2rouge -m `) that you're truncating data (in the preprocessing)

**Comment by yuhui-zh15** [comment]
Hi, it seems you can not reproduce the result as well according to #340? I updated the previous results. I strictly followed the settings mentioned in the OpenNMT-py documents, and I put the scripts I used as follows. Do you think there might be anything wrong?

```sh
python preprocess.py \
    -train_src $CNNDM_DATA_DIR/train.txt.src \
    -train_tgt $CNNDM_DATA_DIR/train.txt.tgt.tagged \
    -valid_src $CNNDM_DATA_DIR/val.txt.src \
    -valid_tgt $CNNDM_DATA_DIR/val.txt.tgt.tagged \
    -save_data $CNNDM_DATA_DIR/CNNDM \
    -src_seq_length 10000 \
    -tgt_seq_length 10000 \
    -src_seq_length_trunc 400 \
    -tgt_seq_length_trunc 100 \
    -dynamic_dict \
    -share_vocab \
    -shard_size 100000
```

```sh
python train.py \
    -save_model models/cnndm \
    -data $CNNDM_DATA_DIR/CNNDM \
    -copy_attn \
    -global_attention mlp \
    -word_vec_size 128 \
    -rnn_size 512 \
    -layers 1 \
    -encoder_type brnn \
    -train_steps 200000 \
    -max_grad_norm 2 \
    -dropout 0. \
    -batch_size 32 \
    -valid_batch_size 32 \
    -optim adagrad \
    -learning_rate 0.15 \
    -adagrad_accumulator_init 0.1 \
    -reuse_copy_attn \
    -copy_loss_by_seqlength \
    -bridge \
    -seed 777 \
    -world_size 1 \
    -gpu_ranks 0 
```

```
python translate.py \
    -gpu 0 \
    -batch_size 20 \
    -beam_size 5 \
    -model models/cnndm_step_170000.pt \
    -src $CNNDM_DATA_DIR/test.txt.src \
    -output testout/cnndm.out \
    -min_length 35 \
    -stepwise_penalty \
    -coverage_penalty summary \
    -beta 5 \
    -length_penalty wu \
    -alpha 0.9 \
    -block_ngram_repeat 3 \
    -ignore_when_blocking "." "</t>" "<t>" 
```

**Comment by pltrdy** [comment]
I mostly worked w/ transformer models, which now comes really close to what has been published.

**Comment by yuhui-zh15** [comment]
Hi, do you have any idea how to reproduce results on NYT dataset? According to the paper _Bottom-up Summarization_, I only need to change `--min_length` to `6` and try different `--alpha` from `0.8` to `1.4` during translation stage. However, I can't reproduce the results and it is far from the results reported in the paper.

```
alpha	R1-R	R1-P	R1-F1	R2-R	R2-P	R2-F1	RL-R	RL-P	RL-F1
0.6	28.52	53.14	32.72	17.87	31.60	20.35	26.22	49.05	30.14	
0.8	31.17	52.12	34.66	19.48	31.43	21.61	28.58	48.02	31.86	
1.0	35.35	49.77	36.99	21.84	30.28	22.95	32.31	45.80	33.93	
1.2	38.83	47.05	38.18	23.76	28.39	23.47	35.45	43.25	34.99	
1.4	40.55	45.27	38.35	24.71	26.94	23.40	37.02	41.58	35.14
paper	-	-	45.13	-	-	30.13	-	-	39.67
```

**Comment by pltrdy** [comment]
No sorry, haven't been working on NYT dataset. 

Not that, in the Bottom Up paper, the NYT scores only concerns (i) `Pointer-Generator` which isn't this repo, but https://github.com/abisee/pointer-generator/ again, model are not the same; (ii) `BottomUp` which is different also, (even if build on top of ONMT-py) see Gehrmann's fork: https://github.com/sebastianGehrmann/bottom-up-summary

**Comment by yuhui-zh15** [comment]
Thank you very much for your help! We will try more ways to reproduce the results.

**Comment by joelowj** [comment]
Hey @pltrdy, I am trying to replicate the results for the Cnndm for transformer using the exact same commands. How Long did you take to train it on a Gpu? Not sure if this is normal, I am already 100k steps into training yet I have not seen improvement in metric since 30k steps. It has been sitting at acc (23) and ppl (300+).

**Comment by pltrdy** [comment]
@joelowj it's not, 300ppl is huge. Note that so-called `steps` really depends on your batch_size, accum_count, and gpu count. For steps to be identical, make sure `batch_size x accum_count x gpu` is the same (in particular, if you have only 1 GPU, double the accum, if the batch_size does not fit into the memory, halve it and double the accum).
